pkg/serving/v1/apply.go

```diff
@@ -0,0 +1,293 @@
+
+	"knative.dev/client/pkg/util"
+)
+
+// Copyright © 2020 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Helper methods supporting Apply()
+
+// patchStrategy is the patch strategy to us when sending an apply patch
+// This is fixed now to a 'normal' JSON patch, as a strategic merge patch is not supported
+// (yet) for a custom resource definition.
+var patchStrategy = types.MergePatchType
+
+func (cl *knServingClient) patch(modifiedService *servingv1.Service, currentService *servingv1.Service, uOriginalService []byte) (bool, error) {
+	uModifiedService, err := getModifiedConfiguration(modifiedService, true)
+	if err != nil {
+		return false, err
+	}
+	hasChanged, err := cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	for i := 1; i <= 5 && apierrors.IsConflict(err); i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		currentService, err = cl.GetService(currentService.Name)
+		if err != nil {
+			return false, err
+		}
+		hasChanged, err = cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	}
+
+	// TODO: Introduce a "force" parameter that would do the same as `kn service create --force`
+	// However, if using a ThreeWayJSONMergePatch, then no conflict can happen, so no need yet for `--force`
+	/*
+		if err != nil && (apierrors.IsConflict(err) || apierrors.IsInvalid(err)) && p.Force {
+			patchBytes, patchService, err = p.deleteAndCreate(current, modified, namespace, name)
+		}
+	*/
+	return hasChanged, err
+}
+
+func (cl *knServingClient) patchSimple(currentService *servingv1.Service, uModifiedService []byte, uOriginalService []byte) (bool, error) {
+	// Serialize the current configuration of the object from the server.
+	uCurrentService, err := encodeService(currentService)
+	if err != nil {
+		return false, err
+	}
+
+	var patch []byte
+	if patchStrategy == types.MergePatchType {
+		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(uOriginalService, uModifiedService, uCurrentService)
+	} else if patchStrategy == types.StrategicMergePatchType {
+		// Compute a three way strategic merge patch to send to server.
+		lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(servingv1.Service{})
+		if err != nil {
+			return false, err
+		}
+		// TODO: Allow "overwrite" to be configured
+		patch, _ = strategicpatch.CreateThreeWayMergePatch(uOriginalService, uModifiedService, uCurrentService, lookupPatchMeta, false)
```

Comment:
ignoring errors....

Reply:
The whole code section is commented out and will never be executed. I will remove that part and find an other way how we can memorize the better alternative way for doing the merge itself.